apps/builder/app/builder/shared/expression-editor.tsx

```diff
@@ -417,7 +421,33 @@ export const ExpressionEditor = ({
         readOnly={readOnly}
         autoFocus={autoFocus}
         value={value}
-        onChange={onChange}
+        onChange={(value) => {
+          try {
+            let hasReplacements = false;
+            // replace unknown webstudio variables with null to prevent invalid compilation
+            const newExpression = validateExpression(value, {
+              effectful: true,
+              transformIdentifier: (identifier) => {
+                if (
+                  decodeDataSourceVariable(identifier) &&
+                  aliases.has(identifier) === false
+                ) {
+                  hasReplacements = true;
+                  return `null`;
```

Comment:
I am not sure about `null`, its shorter, but null means a specific value, where undefined is more like it literally isn't defined, semantically much clearer

Reply:
Also if you imagine that the variable user is trying to access hasn't been assigned a value, the initial value is undefined in js

Reply:
we can also consider using something completely different to not mix up with existing semantics, eg.. `unknown`

Reply:
If you think semantics of undefined are misleading in this case, I think they are great, null doesn't work for me.

Reply:
Btw do we store the id for that insertion if we can't find it immediately in case it appears later? Will it start working when that variable exists in the scope?

Reply:
No, this will not longer store the id. The variable is completely replaced from expression.

I can try to make undefined work. It seems not supported out of the box by jsep.